examples/7guis/cells.rs

```diff
@@ -0,0 +1,245 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+use slint::{Model, ModelRc, SharedString};
+use std::cell::RefCell;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::rc::{Rc, Weak};
+
+slint::slint!(import { MainWindow } from "cells.slint";);
+
+const ROW_COUNT: usize = 100;
+const COL_COUNT: usize = 26;
+
+/// Graph of dependences so that node B depends on node A.
+/// So when node A is updated, node B also need to re updated.
+#[derive(Default)]
+struct DepGraph<A, B> {
+    index1: HashMap<A, HashSet<B>>,
+    index2: HashMap<B, HashSet<A>>,
+}
+
+impl<A: Clone + Eq + Hash, B: Clone + Eq + Hash> DepGraph<A, B> {
+    pub fn add_dep(&mut self, a: A, b: B) {
+        self.index1.entry(a.clone()).or_default().insert(b.clone());
+        self.index2.entry(b).or_default().insert(a);
+    }
+
+    pub fn deps<'a>(&'a self, a: &A) -> impl Iterator<Item = &B> + 'a {
+        self.index1.get(a).into_iter().flat_map(|x| x.iter())
+    }
+
+    pub fn remove_rev_deps(&mut self, b: &B) {
+        if let Some(h) = self.index2.remove(b) {
+            for a in h {
+                self.index1.get_mut(&a).map(|x| x.remove(b));
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum Expr {
+    Value(f32),
+    Cell(usize, usize),
+    Add(Box<(Expr, Expr)>),
+    Sub(Box<(Expr, Expr)>),
+}
+
+fn parse_formula(formula: &str) -> Option<Expr> {
+    let formula = formula.trim();
+    let alpha_n = formula.find(|c: char| !c.is_ascii_alphabetic()).unwrap_or(formula.len());
+    let num_n = formula[alpha_n..]
+        .find(|c: char| !c.is_ascii_digit() && c != '.')
+        .map_or(formula.len(), |x| x + alpha_n);
+
+    let e = if alpha_n > 0 {
+        if alpha_n != 1 {
+            return None;
+        };
+        let col = formula.as_bytes()[0].to_ascii_lowercase() - b'a';
+        let row = formula[alpha_n..num_n].parse().ok()?;
+        Expr::Cell(row, col as usize)
+    } else if num_n > 0 {
+        Expr::Value(formula[alpha_n..num_n].parse().ok()?)
+    } else {
+        return None;
+    };
+
+    let rest = formula[num_n..].trim();
+
+    if rest.is_empty() {
+        Some(e)
+    } else if let Some(x) = rest.strip_prefix("+") {
+        Some(Expr::Add(Box::new((e, parse_formula(x)?))))
+    } else if let Some(x) = rest.strip_prefix("-") {
+        Some(Expr::Sub(Box::new((e, parse_formula(x)?))))
+    } else {
+        None
+    }
+}
+
+#[test]
```

Comment:
Just nitpicking, but I suggest to move that to the very bottom of the file :)

Reply:
I like to have it close to the code it tests.

Reply:
In library code that makes sense, but in this code I think the primary objective of this file is to be read. Some may run this code, a few many copy & paste into their app, but most people will just read this code and try to understand what's happening.

Reply:
and the test help to understand what's hapenning.  and if they copy paste this function, it is good if they copy paste the test as well.

Reply:
Fair point :)